### PR TITLE
Adding Terraform Remote State Store

### PR DESCRIPTION
**Pull request by h2floh**
Adding use of Remote Terraform State Store for real CI/CD pipelines.
Updating Documentaion

Major Change:
`terraform init` is only once executed for all three Terraform Folders. If you loose the remote state information in this folder `InitTerraform.ps1` has to be executed before any other script is called.

Within the scripts there is no `terraform init` in order to avoid passing storage account and resource group name for the blob storage with the state.

Access to the terraform state will be executed without KEY but with the rights of the logged in user.

Fixes also #29

### CODE REVIEW

**Comment by h2floh** [comment]
Creation of Remote Store is failing because AzureCLI connection is type 'ServicePrincipal' and not 'User'. Have to test how to make both possible in `InitTerraform.ps1`

File: Deploy/ActivateSSL.ps1 [hunk 1]
```diff
@@ -41,21 +41,20 @@ param(
...
-$iaCFolder = "IaC"
 $webAppsVariableFile = "$(Get-ScriptPath)/$terraformFolder/webAppVariable.tfvars.json"
 
 # Tell who you are (See HelperFunction.ps1)
 Write-WhoIAm
 
 # 1. Read values from Terraform IaC run (Bot deployment scripts)
 Write-Host "## 1. Read values from Terraform IaC run (Bot deployment scripts)"
-$content = '{ "azure_webApps" : ' + $(terraform output -state="$(Get-ScriptPath)/$iaCFolder/terraform.tfstate" -json webAppAccounts) + '}'
+$content = '{ "azure_webApps" : ' + $(Get-TerraformOutput("webAppAccounts")) + '}'
```
Comment (angie4u) [inline Deploy/ActivateSSL.ps1:51]: It is just question! Now do we store tfstates under webapp which is the place bot deployed as well? Isn't there any security concern cause webapp is publicly opened?

File: Deploy/IaC/bot.tf [hunk 2]
```diff
@@ -165,10 +165,7 @@ resource "azurerm_cognitive_account" "LUISAuthoring" {
   resource_group_name = azurerm_resource_group.GeoBot.name
   kind                = "LUIS.Authoring"
 
-  sku {
-    name = "F0" // S0 not available
-    tier = "Free" // Standard not available
-  }
+  sku_name = "F0" // S0 not available
```
Comment (angie4u) [inline Deploy/IaC/bot.tf:168]: It is just a question. Is it a default setting to deploy or is it not possible to deploy with S0 sku?

File: Deploy/InitTerraform.ps1 [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+# Get info around current connection
+$currentConnection = az account show | ConvertFrom-Json
+$success = $success -and $?
+
+# Save Principal Type and Id
+$principalType = $currentConnection.user.type
+$principalId = ""
+
+if($principalType -eq "user")
```
Comment (angie4u) [inline Deploy/InitTerraform.ps1:77]: Does the behavior differ depending on what the 'service principal type' is? Or is there only difference command which should be executed according to 'service principal type'?

File: Deploy/IaC/bot.tf (hunk 2 above)
Comment (h2floh) [reply Deploy/IaC/bot.tf:168]: Turns out that the **Authoring REST** calls are free of charge and therefore can only be deployed with SKU F0. In contrast to the query **Endpoint REST** calls. They can be created in any available SKU, there is only F0 and S0 available as I know. 
https://github.com/h2floh/GeoDistributedAzureBot/blob/4cef20af17ec63a43dcf1934bb48c8688515e7df/Deploy/IaC/regions.tf#L130

File: Deploy/ActivateSSL.ps1 (hunk 1 above)
Comment (h2floh) [reply Deploy/ActivateSSL.ps1:51]: No, we store the Terraform State in a separate storage account created in the new InitTerraform.ps1 script https://github.com/h2floh/GeoDistributedAzureBot/blob/420458620b71837701d7b6bf5813a62977dfa262/Deploy/InitTerraform.ps1#L66

 The above code will retrieve the Terraform Output details for the WebApps, from the initial run from DeployInfrastructure.ps1, the info is stored as Terraform State file in the Storage Account.

File: Deploy/InitTerraform.ps1 (hunk 3 above)
Comment (h2floh) [reply Deploy/InitTerraform.ps1:77]: You can log into AzureCLI as a user or with a service principal. We need e.g. the service principal login for our test pipeline.
In order to give access rights on the Storage Account for the Terraform State Store we need the AAD ObjectId. However the way how we can retrieve the 'ObjectId' of a **user** or a **service principal** is different. So I try to distinct what login type is used and retrieve the objectId accordingly.

**Review by angie4u — APPROVED** [review]
It looks good to me! Thank you for the hard work! :)

**Pull request merged**